carto/auth.py

```diff
@@ -26,7 +26,43 @@
     from urlparse import urlparse
 
 
-class APIKeyAuthClient(BaseAuthClient):
+class _UsernameGetter:
+    def get_user_name(self, base_url):
+        try:
+            url_info = urlparse(base_url)
+            # On-Prem:
+            #   /user/<username>
+            m = re.search('^/user/([^/]+)/.*$', url_info.path)
```

Comment:
Usually, best practice is to compile the regex: https://docs.python.org/2/library/re.html#re.compile (and maybe assign the compiled object to a constant). If you use it once, there is no penalty. And if you use it multiple times, you are saving the compile team in each successive usage.

Not sure how many times this is going to get called though, if it's typically just once, it may not be worth it.

Reply:
Keep in mind that you're :trollface:-ing other people here ;-) TBH, I prefer keeping changes to the bare minimum in this PR. I just extracted the mixing to avoid duplication, improving it is beyond the scope.